crates/biome_js_analyze/src/lint/correctness/no_undeclared_dependencies.rs

```diff
@@ -34,32 +38,199 @@ declare_lint_rule! {
     /// ```js,ignore
     /// import assert from "node:assert";
     /// ```
+    ///
+    /// ## Options
+    ///
+    /// This rule supports the following options:
+    /// - `devDependencies`: If set to `false`, then the rule will show an error when `devDependencies` are imported. Defaults to `true`.
+    /// - `peerDependencies`: If set to `false`, then the rule will show an error when `peerDependencies` are imported. Defaults to `true`.
+    /// - `optionalDependencies`: If set to `false`, then the rule will show an error when `optionalDependencies` are imported. Defaults to `true`.
+    ///
+    /// You can set the options like this:
+    /// ```json
+    /// {
+    ///   "options": {
+    ///     "devDependencies": false,
+    ///     "peerDependencies": false,
+    ///     "optionalDependencies": false
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// You can also use an array of globs instead of literal booleans:
+    /// ```json
+    /// {
+    ///   "options": {
+    ///     "devDependencies": ["**/*.test.js", "**/*.spec.js"]
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// When using an array of globs, the setting will be set to `true` (no errors reported)
+    /// if the name of the file being linted (i.e. not the imported file/module) matches a single glob
+    /// in the array, and `false` otherwise.
     pub NoUndeclaredDependencies {
         version: "1.6.0",
         name: "noUndeclaredDependencies",
         language: "js",
+        sources: &[
+            RuleSource::EslintImport("no-extraneous-dependencies"),
+        ],
         recommended: false,
     }
 }
 
+#[derive(Clone, Debug, Default, serde::Deserialize, serde::Serialize)]
+enum DependencyAvailability {
+    /// Dependencies are always available.
+    #[default]
+    Available,
+
+    /// Dependencies are never available.
+    Unavailable,
+
+    /// Dependencies are available in files that matches any of the globs.
+    FilesGlob(Box<[RestrictedGlob]>),
+}
+
+impl PartialEq for DependencyAvailability {
+    fn eq(&self, other: &Self) -> bool {
+        match self {
+            Self::Available => matches!(other, Self::Available),
+            Self::Unavailable => matches!(other, Self::Unavailable),
+            Self::FilesGlob(a) => match other {
+                Self::FilesGlob(b) => a
+                    .iter()
+                    .zip(b.iter())
```

Comment:
Be careful using `zip` like this, since it terminates at the shortest list. So this comparison will still return `true` even if one of the lists has excess items.

Reply:
Added length check before the zip: [5b57e80](https://github.com/biomejs/biome/pull/4376/commits/5b57e80c9c0a5e0d59365e43a4eb096ea4544fbd)